pyinstaller/pyinstaller
# MacOS: Make apps accept both drag/drop & URL events via ARGV EMU before startup. After startup: Apps now accept URL & drag/drop events via Apple Event forwarding

Hi,

Electron Cash developer (and occasional Electrum contributor) here!  We use pyinstaller for our app and love it.

However we noticed that the bootloader parent process wasn't properly picking up the MacOS "URL" events (kAEGetURL as it's called on macOS) and it wasn't forwarding them to the child process.  

It also didn't accept drag/drops of files after the program was started.

There was a patch that someone wrote a couple of years back that supposedly fixed this -- but it didn't actually work at all as intended on modern MacOS. I tried on El Capitan and up. No luck. 

So-- URI/URL handling wasn't working at all.  Not at startup, and not if the app was already running.

Drag/drop only worked previously if the app wasn't running.  After startup -- the events wouldn't make it to the GUI framework.

So I went ahead and fixed it by rewriting some of the bootloader code that was in the aforementioned commit.  Note that this code only gets compiled in if using WINDOWED building mode on macOS.

### To Test ###

Create an app with pyinstaller on Mac that accepts custom URI/URLs for a test scheme, say `myscheme:`, by adding the following XML to your `Info.plist`, using the `plutil` utility from the Terminal:

```plutil -insert 'CFBundleURLTypes' -xml '<array><dict> <key>CFBundleURLName</key> <string>myscheme</string> <key>CFBundleURLSchemes</key> <array><string>myscheme</string></array> </dict></array>' -- /path/to/YOUR.app/Contents/Info.plist```

Once that's done, and the app has the new `myscheme:` URL support (be sure your app was coded up to accept URLs via whatever GUI framework you are using.  On PyQt this would be QFileOpenEvent in your QApplication event() function).

Next, try and open up `myscheme:` style URLs:

1. From Terminal: `open myscheme:this_is_a_test?foo=bar`
2. From a browser: (in the location bar) `myscheme:this_is_a_test?foo=bar`

Your app should receive that URL from macOS (if you are using PyQt it comes in as a QFileOpenEvent). If the app wasn't started, it should be started and receive that URL.

However, that wasn't working previously.

But with this patch, everything works.  If the app wasn't started, the URL is appended to the sys.argv of the python script.  If the app was running, an AppleEvent is forwarded encapsulating the URL request.

Let me know if you have questions or if I can help test!

We would like very much to have this merged into pyinstaller for a future release so that our app can be built by anyone and work correctly! 

Thanks!
